### PR TITLE
Add MAP_SHARED option for torch.load(mmap=True)

### LINKED ISSUE #124528: Add option to `torch.load(mmap=True)` to do `MAP_SHARED` rather than `MAP_PRIVATE`
### 🚀 The feature, motivation and pitch

Currently `torch.load(mmap=True)` calls `mmap(2)` with `MAP_PRIVATE`, this ensures that changes to the loaded checkpoint are not written back to the file.

As suggested by @albanD there might be a way to do postprocessing on checkpoints in CPU RAM-constrained environments if we  pass `MAP_SHARED` instead 

This would just involve passing `shared=True` to `UntypedStorage.from_file`

### Alternatives

_No response_

### Additional context

_No response_

cc @mruberry

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Fixes #124528

Going over the options for our MapAllocator and what they do, I don't think any other of them need to be piped up to `torch.load`

https://github.com/pytorch/pytorch/blob/4f29103749c5011529f1abb10b1508a682588909/aten/src/ATen/MapAllocator.h#L8-L16

~However, I wonder if this `MmapVisibility(Enum)` is a good way to represent "or-ing" together of `mmap` flags if we want to extend it in the future. I looked over the flags for [`mmap(2)`](https://man7.org/linux/man-pages/man2/mmap.2.html), and could not immediately see how most of them would be useful for `torch.load` (would maybe `MAP_LOCKED` (like `mlock`) or `MAP_HUGE` ever be worthwhile?)~

Using the flags provided by the python `mmap` library so that we can extend the allowed flags and pipe them down to the cpp `mmap` call if there is a need for other flags in the future

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124891
* __->__ #124889

cc @albanD

### CODE REVIEW

**Comment by vadimkantorov** [comment]
btw, is it possible to create a mmap-backed tensor without torch.save/load?

**Comment by mikaylagawarecki** [comment]
@vadimkantorov [`torch.from_file`](https://pytorch.org/docs/stable/generated/torch.from_file.html) might be what you are looking for, note that it creates a flat 1D tensor so the output would need to be reshaped

File: torch/serialization.py [hunk 1]
```diff
@@ -105,6 +106,36 @@ def set_default_load_endianness(endianness):
...
+        default_mmap_options: int
+    '''
+    return _default_mmap_options
+
+def set_default_mmap_options(flags: int):
+    '''
+    Set default mmap options for :func:`torch.load` with ``mmap=True`` to flags.
+
+    For now, only either ``mmap.MAP_PRIVATE`` or ``mmap.MAP_SHARED`` are supported,
+    in the future we might extend this to support more options.
```
Comment (albanD) [inline torch/serialization.py:128]: nit: please open an issue if you need any other option to be added here.

**Review by albanD — APPROVED** [review]
nit on doc but sounds great otherwise!

**Pull request closed without merging**